src/bio/bio_xstream.c

```diff
@@ -1219,25 +1291,30 @@ init_blobstore_ctxt(struct bio_xs_context *ctxt, int tgt_id)
 void
 bio_xsctxt_free(struct bio_xs_context *ctxt)
 {
-	int	rc = 0;
+	int			 rc = 0;
+	enum smd_type		 st;
+	struct bio_xs_blobstore	*bxb;
 
 	/* NVMe context setup was skipped */
 	if (ctxt == NULL)
 		return;
 
 	ctxt->bxc_ready = 0;
-	if (ctxt->bxc_io_channel != NULL) {
-		spdk_bs_free_io_channel(ctxt->bxc_io_channel);
-		ctxt->bxc_io_channel = NULL;
-	}
-
-	if (ctxt->bxc_blobstore != NULL) {
-		put_bio_blobstore(ctxt->bxc_blobstore, ctxt);
-
-		if (is_bbs_owner(ctxt, ctxt->bxc_blobstore))
-			bio_fini_health_monitoring(ctxt->bxc_blobstore);
+	for (st = SMD_TYPE_DATA; st < SMD_TYPE_MAX; st++) {
+		bxb = ctxt->bxc_xs_blobstores[st];
+		if (bxb == NULL)
+			continue;
 
-		ctxt->bxc_blobstore = NULL;
+		if (bxb->bxb_io_channel != NULL) {
+			spdk_bs_free_io_channel(bxb->bxb_io_channel);
+			bxb->bxb_io_channel = NULL;
+		}
+		if (bxb->bxb_blobstore != NULL) {
+			bio_blobstore_free(bxb, ctxt);
+			bxb->bxb_blobstore = NULL;
```

Comment:
Owner xstream, setup, shutdown procedures need be carefully considered to avoid deadlock. Please be aware that blobstore is always load and unload by owner xstream, and unloading blobstore needs to wait for blobstore users (other xstreams sharing the same blobstore) drop their reference first.

Reply:
Indeed, deadlock might happen in the following case:
1: xstream1                             xstrem2
put blobstore1                         put blobstore2
put blobstore2                         put blobstore1

blobstore1 && blobstore2 are shared by xstream1 and xstream2.
xstream1 will wait xstream2 to free blob1, while xstream2 need wait xstream1 to free blobstore2, deadlock happen,
we might fix this problem by putting blobs in the same order.

Reply:
As discussed this won't have problem, because now we support following three cases:

1. meta && data & wal blobstore share same device, which won't have problem as before.
2. metat && wal share same blobstore, data use seperate blobstore
3. meta, wal, data blobstore use three different devices.

For the case2 and 3, since we don't support one device will have different types, it won't have problems for now.